src/keys/bip39.rs

```diff
@@ -5,24 +5,132 @@
 
 // https://doc.rust-lang.org/std/primitive.str.html
 // "String slices are always valid UTF-8."
-type Mnemonic = str;
-type Passphrase = str;
-type Seed = [u8; 64];
 
 use alloc::string::{String, ToString};
+use core::convert::TryFrom;
+use core::ops::Deref;
 
-use unicode_normalization::UnicodeNormalization;
+use unicode_normalization::{is_nfkd, UnicodeNormalization};
+use zeroize::{Zeroize, ZeroizeOnDrop, Zeroizing};
 
-pub fn mnemonic_to_seed(m: &Mnemonic, p: &Passphrase, s: &mut Seed) {
-    let m = m.chars().nfkd().collect::<String>();
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+    InvalidEntropyCount(usize),
+    NoSuchWord(String),
+    ChecksumMismatch,
+    UnnormalizedMnemonic,
+    UnnormalizedPassphrase,
+    BadWordlist,
+    BadSeparator,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct MnemonicRef<'a>(&'a str);
+
+impl<'a> Deref for MnemonicRef<'a> {
+    type Target = str;
+    fn deref(&self) -> &str {
+        self.0
+    }
+}
+
+impl<'a> TryFrom<&'a str> for MnemonicRef<'a> {
+    type Error = Error;
+    fn try_from(mnemonic_str: &'a str) -> Result<Self, Error> {
+        if is_nfkd(mnemonic_str) {
+            Ok(MnemonicRef(mnemonic_str))
+        } else {
+            Err(Error::UnnormalizedMnemonic)
+        }
+    }
+}
+
+impl<'a> PartialEq<str> for MnemonicRef<'a> {
+    fn eq(&self, other: &str) -> bool {
+        self.0.eq(other)
+    }
+}
+
+#[derive(ZeroizeOnDrop)]
```

Comment:
Can we add Clone here?

Reply:
I'd say it's ok for tests to clone strings. But added derive Clone anyway.